### PR TITLE
adding move (mv) and rename commands with docs

### LINKED ISSUE #65: [mv] add move command to move images to (non traditional) locations
It should be possible a user to do:

```
sregistry mv container/uri <path>
```
The user should provide a URI, and path can be either a container name, or a folder. The logic will be the following:

1. If it's a directory that exists, then move the image there, maintaining the original name
2. If it's not a directory and doesn't exist, ensure the dirname of it does. Then name the image what the user asks.

Then, the container will be (physically moved) and the database updated.

Additionally, all commands (if this isn't already done) should do a check that an image exists before doing anything, and if not, remove the record from the database and tell the user.

### PR DESCRIPTION
**Pull request by vsoch**
This PR will close #65 and upon merge, be pushed to pypi as version 0.0.67. @shahzebsiddiqui as you requested this feature I'd like you to test and see if it is to your liking before merge. To be clear, you will need to clone this branch and install from it - there is no package on pypi yet, and the change only exists here.

```
git clone -b add/move https://www.github.com/singularityhub/sregistry-cli
cd sregistry-cli
python setup.py install
``` 
Would clone a fresh repo. You can also follow the instructions below to check out a branch then proceed with install.

Please see docs and instructions in https://github.com/singularityhub/sregistry-cli/blob/add/move/docs/getting-started/commands.md#move

### CODE REVIEW

**Comment by vsoch** [comment]
This will also close https://github.com/singularityhub/sregistry-cli/issues/68

**Comment by vsoch** [comment]
Added another fix to clean up build folders in /tmp to close #69

**Comment by vsoch** [comment]
@shahzebsiddiqui I just added a fix to the url not defined variable that you were seeing for the dummy, please re-test!

**Comment by vsoch** [comment]
ping @shahzebsiddiqui this is waiting for your review!

**Comment by shahzebsiddiqui** [comment]
Okay I will try to do it tomorrow, sorry I was busy today with meetings

**Comment by shahzebsiddiqui** [comment]
One minor thing when doing installation I got a syntax error after pulling your changes

I did a new clone any my HEAD is at 
```

commit b9e60ca1f626ab2fd0ceec2b9f4cc11abd621d41
Author: vsoch <vsochat@stanford.edu>
Date:   Sat Feb 3 09:50:18 2018 -0800

    typo


creating /nfs/grid/software/RHEL7/easybuild/software/Core/Anaconda3/4.2.0/lib/python3.5/site-packages/sregistry-0.0.67-py3.5.egg
Extracting sregistry-0.0.67-py3.5.egg to /nfs/grid/software/RHEL7/easybuild/software/Core/Anaconda3/4.2.0/lib/python3.5/site-packages
  File "/nfs/grid/software/RHEL7/easybuild/software/Core/Anaconda3/4.2.0/lib/python3.5/site-packages/sregistry-0.0.67-py3.5.egg/sregistry/database/dummy.py", line 56
    def __init__(self, image_path, client_name, url, names)
                                                          ^
SyntaxError: invalid syntax

sregistry 0.0.67 is already the active version in easy-install.pth
Installing sregistry script to /nfs/grid/software/RHEL7/easybuild/software/Core/Anaconda3/4.2.0/bin

Installed /nfs/grid/software/RHEL7/easybuild/software/Core/Anaconda3/4.2.0/lib/python3.5/site-packages/sregistry-0.0.67-py3.5.egg
Processing dependencies for sregistry==0.0.67
Searching for Pygments==2.1.3
Best match: Pygments 2.1.3
Adding Pygments 2.1.3 to easy-install.pth file
Installing pygmentize script to /nfs/grid/software/RHEL7/easybuild/software/Core/Anaconda3/4.2.0/bin

Using /nfs/grid/software/RHEL7/easybuild/software/Core/Anaconda3/4.2.0/lib/python3.5/site-packages
Searching for requests==2.18.4
Best match: requests 2.18.4
Adding requests 2.18.4 to easy-install.pth file

Using /nfs/grid/software/RHEL7/easybuild/software/Core/Anaconda3/4.2.0/lib/python3.5/site-packages
Searching for urllib3==1.22
Best match: urllib3 1.22
Adding urllib3 1.22 to easy-install.pth file

Using /nfs/grid/software/RHEL7/easybuild/software/Core/Anaconda3/4.2.0/lib/python3.5/site-packages
Searching for chardet==3.0.4
Best match: chardet 3.0.4
Adding chardet 3.0.4 to easy-install.pth file
Installing chardetect script to /nfs/grid/software/RHEL7/easybuild/software/Core/Anaconda3/4.2.0/bin

Using /nfs/grid/software/RHEL7/easybuild/software/Core/Anaconda3/4.2.0/lib/python3.5/site-packages
Searching for idna==2.6
Best match: idna 2.6
Adding idna 2.6 to easy-install.pth file

Using /nfs/grid/software/RHEL7/easybuild/software/Core/Anaconda3/4.2.0/lib/python3.5/site-packages
Searching for certifi==2018.1.18
Best match: certifi 2018.1.18
Adding certifi 2018.1.18 to easy-install.pth file

Using /nfs/grid/software/RHEL7/easybuild/software/Core/Anaconda3/4.2.0/lib/python3.5/site-packages
Finished processing dependencies for sregistry==0.0.67

```

**Comment by vsoch** [comment]
Not sure how I missed that ! I just pushed a fix. Could you try again?

**Comment by shahzebsiddiqui** [comment]
So I tested your changes it looks better but I observed something strange. After moving an image and then doing **sregistry rmi** the image metadata was still present, but when doing with commit id then the image was removed successfully.

```
-bash-4.2$ sregistry pull shahzebsiddiqui/eb-singularity:centos-7.3.1611
[client|hub] [database|sqlite:////home/siddis14/.singularity/sregistry.db]
Progress |===================================| 100.0%
[container][new] shahzebsiddiqui/eb-singularity:centos-7.3.1611@19a949316fb402756b70956dd89464b9
Success! /home/siddis14/.singularity/shub/shahzebsiddiqui-eb-singularity:centos-7.3.1611@19a949316fb402756b70956dd89464b9.simg

-bash-4.2$ sregistry images
Containers:   [date]   [location]  [client]     [uri]
1  February 06, 2018    local      [hub]        shahzebsiddiqui/eb-singularity:centos-7.3.1611@19a949316fb402756b70956dd89464b9

-bash-4.2$ sregistry get shahzebsiddiqui/eb-singularity:centos-7.3.1611
/home/siddis14/.singularity/shub/shahzebsiddiqui-eb-singularity:centos-7.3.1611@19a949316fb402756b70956dd89464b9.simg

-bash-4.2$ sregistry mv shahzebsiddiqui/eb-singularity:centos-7.3.1611 /tmp
[client|hub] [database|sqlite:////home/siddis14/.singularity/sregistry.db]
[mv] shahzebsiddiqui/eb-singularity:centos-7.3.1611 => /tmp/shahzebsiddiqui-eb-singularity:centos-7.3.1611@19a949316fb402756b70956dd89464b9.simg

-bash-4.2$ sregistry get shahzebsiddiqui/eb-singularity:centos-7.3.1611
/tmp/shahzebsiddiqui-eb-singularity:centos-7.3.1611@19a949316fb402756b70956dd89464b9.simg

-bash-4.2$ sregistry rmi /tmp/shahzebsiddiqui-eb-singularity:centos-7.3.1611
[client|hub] [database|sqlite:////home/siddis14/.singularity/sregistry.db]

-bash-4.2$ sregistry images
Containers:   [date]   [location]  [client]     [uri]
1  February 06, 2018    local      [hub]        shahzebsiddiqui/eb-singularity:centos-7.3.1611@19a949316fb402756b70956dd89464b9

-bash-4.2$ sregistry rmi shahzebsiddiqui/eb-singularity:centos-7.3.1611@19a949316fb402756b70956dd89464b9

[client|hub] [database|sqlite:////home/siddis14/.singularity/sregistry.db]
/tmp/shahzebsiddiqui-eb-singularity:centos-7.3.1611@19a949316fb402756b70956dd89464b9.simg
[rmi] /tmp/shahzebsiddiqui-eb-singularity:centos-7.3.1611@19a949316fb402756b70956dd89464b9.simg
```

The test below I tried pulling the same image twice, I am not sure if sregistry is meant to pull image from remote if you have pulled the image already. Does sregistry check the database before pulling the image or it is doing it anyways in case the remote image has changed. It may be the case that remote image has been updated with the same tag but the commit tag will be different. But then again, not sure if user does a **git commit --amend** and  then do a force rebuild in SHUB the container image can change while preserving the same commit tag. This may be another issue, but I dont think that is good practice and can be resolved.

```
-bash-4.2$ sregistry pull  shahzebsiddiqui/eb-singularity:centos-7.3.1611
[client|hub] [database|sqlite:////home/siddis14/.singularity/sregistry.db]
Progress |===================================| 100.0%
[container][new] shahzebsiddiqui/eb-singularity:centos-7.3.1611@19a949316fb402756b70956dd89464b9
Success! /home/siddis14/.singularity/shub/shahzebsiddiqui-eb-singularity:centos-7.3.1611@19a949316fb402756b70956dd89464b9.simg
-bash-4.2$ sregistry images
Containers:   [date]   [location]  [client]     [uri]
1  February 06, 2018    local      [hub]        shahzebsiddiqui/eb-singularity:centos-7.3.1611@19a949316fb402756b70956dd89464b9
-bash-4.2$ sregistry pull  shahzebsiddiqui/eb-singularity:centos-7.3.1611
[client|hub] [database|sqlite:////home/siddis14/.singularity/sregistry.db]
Progress |===================================| 100.0%
[container][update] shahzebsiddiqui/eb-singularity:centos-7.3.1611@19a949316fb402756b70956dd89464b9
Success! /home/siddis14/.singularity/shub/shahzebsiddiqui-eb-singularity:centos-7.3.1611@19a949316fb402756b70956dd89464b9.simg
-bash-4.2$ sregistry images
Containers:   [date]   [location]  [client]     [uri]
1  February 06, 2018    local      [hub]        shahzebsiddiqui/eb-singularity:centos-7.3.1611@19a949316fb402756b70956dd89464b9
```

**Comment by vsoch** [comment]
That is nothing new - it's the current behavior. It means that you are keeping a general metadata record (the one without a version) from the registry, and that is different from the image. if you think this should work differently, file another (different) issue, it's not related to this.

**Comment by shahzebsiddiqui** [comment]
@vsoch just updated the thread.

**Comment by vsoch** [comment]
please open a new issue to isolate from this pull request, thanks! And here please continue (or finish and approve) the PR so it can be merged (or updated).

**Comment by vsoch** [comment]
oh I see the other comments for move - yes I will address these first, likely later today, I need to take a break for a bit.

**Comment by vsoch** [comment]
ah ok! For the first part it's a bug - the user shouldn't be able to "move" a remote image because.. they don't have it. Let me fix that first.

**Comment by vsoch** [comment]
okay for the second where it adds a `/`, this is due to not exiting when it should, now fixed.

**Comment by vsoch** [comment]
working on the permission issue now!

**Comment by shahzebsiddiqui** [comment]
I think we are good to close right?

**Comment by vsoch** [comment]
@shahzebsiddiqui , for the mv command if you just specify `myimage.simg`, this would suggest moving to the $PWD, and this is what move will do. If you would want to rename the image file in the storage, then we can provide another entry point for this. You can open an issue and we can add in another PR.

**Comment by shahzebsiddiqui** [comment]
Okay i will create an issue

**Comment by vsoch** [comment]
>> The test below I tried pulling the same image twice, and yes you found a bug! I think I just fixed:

```
sregistry pull vsoch/hello-world
[client|hub] [database|sqlite:////home/vanessa/.singularity/sregistry.db]
Progress |===================================| 100.0% 
[container][new] vsoch/hello-world:latest@ed9755a0871f04db3e14971bec56a33f
Success! /home/vanessa/.singularity/shub/vsoch-hello-world:latest@ed9755a0871f04db3e14971bec56a33f.simg

$ sregistry pull vsoch/hello-world
[client|hub] [database|sqlite:////home/vanessa/.singularity/sregistry.db]
ERROR Image exists! Remove first, or use --force to overwrite

$ sregistry rmi vsoch/hello-world
[client|hub] [database|sqlite:////home/vanessa/.singularity/sregistry.db]
/home/vanessa/.singularity/shub/vsoch-hello-world:latest@ed9755a0871f04db3e14971bec56a33f.simg
[rmi] /home/vanessa/.singularity/shub/vsoch-hello-world:latest@ed9755a0871f04db3e14971bec56a33f.simg

$ sregistry pull vsoch/hello-world
[client|hub] [database|sqlite:////home/vanessa/.singularity/sregistry.db]
Progress |===================================| 100.0% 
[container][new] vsoch/hello-world:latest@ed9755a0871f04db3e14971bec56a33f
Success! /home/vanessa/.singularity/shub/vsoch-hello-world:latest@ed9755a0871f04db3e14971bec56a33f.simg
```

>> Does sregistry check the database before pulling the image or it is doing it anyways in case the remote image has changed. 

It checks based on the version in the manifest! If the image file version is the same, it won't re-pull.

>> It may be the case that remote image has been updated with the same tag but the commit tag will be different. 

If the image uri that is parsed from the manifest is the same, it is considered the same.

>> But then again, not sure if user does a git commit --amend and then do a force rebuild in SHUB the container image can change while preserving the same commit tag. This may be another issue, but I dont think that is good practice and can be resolved.

It is up to the user how to manage building and committing.

**Comment by vsoch** [comment]
I have finished addressing your issues, please test the current PR, thank you!

**Comment by shahzebsiddiqui** [comment]
Looks good to me
```
[root@adwnode38 sregistry-cli]# sregistry pull vsoch/hello-world
[client|hub] [database|sqlite:////root/.singularity/sregistry.db]
Progress |===================================| 100.0%
[container][new] vsoch/hello-world:latest@ed9755a0871f04db3e14971bec56a33f
Success! /root/.singularity/shub/vsoch-hello-world:latest@ed9755a0871f04db3e14971bec56a33f.simg
[root@adwnode38 sregistry-cli]# sregistry images
Containers:   [date]   [location]  [client]     [uri]
1  February 06, 2018    local      [hub]        vsoch/hello-world:latest@ed9755a0871f04db3e14971bec56a33f

[root@adwnode38 sregistry-cli]# sregistry pull vsoch/hello-world
[client|hub] [database|sqlite:////root/.singularity/sregistry.db]
ERROR Image exists! Remove first, or use --force to overwrite

[root@adwnode38 sregistry-cli]# sregistry rmi vsoch/hello-world
[client|hub] [database|sqlite:////root/.singularity/sregistry.db]
/root/.singularity/shub/vsoch-hello-world:latest@ed9755a0871f04db3e14971bec56a33f.simg
[rmi] /root/.singularity/shub/vsoch-hello-world:latest@ed9755a0871f04db3e14971bec56a33f.simg

[root@adwnode38 sregistry-cli]# sregistry pull vsoch/hello-world
[client|hub] [database|sqlite:////root/.singularity/sregistry.db]
Progress |===================================| 100.0%
[container][new] vsoch/hello-world:latest@ed9755a0871f04db3e14971bec56a33f
Success! /root/.singularity/shub/vsoch-hello-world:latest@ed9755a0871f04db3e14971bec56a33f.simg

```

Though sregistry pull --name is not working as expected

```
[root@adwnode38 sregistry-cli]# sregistry pull --name helloworld vsoch/hello-world
[client|hub] [database|sqlite:////root/.singularity/sregistry.db]
Progress |===================================| 100.0%
[container][new] vsoch/hello-world:latest@ed9755a0871f04db3e14971bec56a33f
Success! /root/.singularity/shub/vsoch-hello-world:latest@ed9755a0871f04db3e14971bec56a33f.simg
[root@adwnode38 sregistry-cli]# sregistry images
Containers:   [date]   [location]  [client]     [uri]
1  February 06, 2018    local      [hub]        vsoch/hello-world:latest@ed9755a0871f04db3e14971bec56a33f
[root@adwnode38 sregistry-cli]# sregistry get vsoch/hello-world
/root/.singularity/shub/vsoch-hello-world:latest@ed9755a0871f04db3e14971bec56a33f.simg
```

Tried pulling image again using --force --name

```
[root@adwnode38 sregistry-cli]# sregistry pull vsoch/hello-world --force --name helloworld2
[client|hub] [database|sqlite:////root/.singularity/sregistry.db]
Progress |===================================| 100.0%
[container][update] vsoch/hello-world:latest@ed9755a0871f04db3e14971bec56a33f
Success! /root/.singularity/shub/vsoch-hello-world:latest@ed9755a0871f04db3e14971bec56a33f.simg
[root@adwnode38 sregistry-cli]# sregistry images
Containers:   [date]   [location]  [client]     [uri]
1  February 06, 2018    local      [hub]        vsoch/hello-world:latest@ed9755a0871f04db3e14971bec56a33f

```

**Comment by vsoch** [comment]
First, I have added the ability to rename an image, the difference between this and move, and examples, shown here:

https://github.com/singularityhub/sregistry-cli/blob/1396e32ad0d87e753d49c1eea3a8f3f41a603627/docs/getting-started/commands.md#rename

TLDR: if the user wants to rename an image in its current storage folder, use "rename." Lack of specifying a directory assumes the place where it's being stored. If the user wants to actually change the folder of storage (and not having a directory means present working directory) then move is used.

This addition will resolve the issues brought up by @shahzebsiddiqui in slack.

 1. Pull a remote image and rename it
 2. Move an image to another directory
 3. Rename an existing image

And the updates to the comments above:

 - a custom name should be honored.
Before this wasn't happening - the name was being ignored. Here we honor it.

```
sregistry pull --name tacos.simg vsoch/hello-world
[client|hub] [database|sqlite:////home/vanessa/.singularity/sregistry.db]
Progress |===================================| 100.0% 
[container][update] vsoch/hello-world:latest@ed9755a0871f04db3e14971bec56a33f
Success! /home/vanessa/tacos.simg
```

@shahzebsiddiqui this should hopefully be the last review, unless there are other issues! When you get a chance, please test once more!

**Comment by vsoch** [comment]
Thanks for your help @shahzebsiddiqui !

**Comment by shahzebsiddiqui** [comment]
my pleasure 👍 good work
